### DISCUSSION TITLE
How to build node project within a Dynamic Resource Provider?

### DISCUSSION THREAD

**Discussion by gjolund** [pulumi/pulumi#7310] [Pulumi in Practice]
Hello,

This may come across more as a question about best practices than a bug report. I apologize if this is the wrong venue.

Issue:

I am trying to use Pulumi to build a static SPA (ts react app) and deploy it to S3.

I have been able to find lots of documentation on how to deploy static assets to S3 via Pulumi, and everything there is working fine.

Where I am stuck is at the build step.

My SPA has dependencies on environment variables that are defined by Pulumi (API url for example), that need to be provided to my SPA at build time.

Currently this is what I have:

```typescript
function buildWeb(modulePath: string, env: NodeJS.ProcessEnv = {}) {
  // Yarn doesn't have a programmatic api https://github.com/yarnpkg/yarn/issues/906
  // spawn.sync doesn't know how to resolve modules in a pulumi serialized function
  // https://www.pulumi.com/docs/tutorials/aws/serializing-functions/#serializing-javascript-functions
  const command = require.resolve("yarn/bin/yarn");
  const args = ["--cwd", modulePath, "build"];
  // spawn.sync is provided by the cross-spawn package https://www.npmjs.com/package/cross-spawn.
  const proc = spawn.sync(command, args, { env });
  if (proc.status === 0) {
    console.error(`\`${command} ${args.join(" ")}\` success`);
    console.log("stdout", proc.stdout.toString());
  } else {
    console.error(`\`${command} ${args.join(" ")}\` failed`);
    console.log("stderr", proc.stderr.toString());
  }
}
```

This function is being called within a dynamic resource provider that is deploying the build static assets to an S3 bucket. The provider is working fine, and if I build the app outside of Pulumi everything works great.

However if I run the `buildWeb` function above within the dynamic resource provider I get the following error.

```sh
stderr Yarn requires Node.js 4.0 or higher to be installed.
```

In summary I have 3 Question:

1. Is this use case supported by Pulumi and documented elsewhere (building compiled assets).
2. Why does it seem like node is unavailable in the spawned build process?
3. What is the best way to handle errors in Pulumi providers?

**Comment by komalali** [comment]
@jaxxstorm do you know the answers to these questions?

**Comment by gjolund** [comment]
@komalali it seems like question 2 works within a github action, so it seems like the node version error may just be a local issue.
